test/Atmos/EDMF/edmf_model.jl

```diff
@@ -0,0 +1,284 @@
+    ω_pr::FT = 53.0 / 13.0
+    "Prandtl number scale"
+    Pr_n::FT = 1 #0.74
+    "Critical Richardson number"
+    Ri_c::FT = 0.25
+    "Random small number variabile that should be adressed"
+    random_minval::FT = 1e-9
+    "smooth minimum's' fractional upper bound"
+    smin_ub::FT = 0.1
+    "smooth minimum's regularization minimum"
+    smin_rm::FT = 1.5
+
+end
+
+abstract type AbstractStatisticalModel end
+struct SubdomainMean <: AbstractStatisticalModel end
+struct GaussQuad <: AbstractStatisticalModel end
+struct LogNormalQuad <: AbstractStatisticalModel end
+
+"""
+    MicrophysicsModel
+A `MicrophysicsModel` for EDMF, containing
+all related free parameters and assumed subdomain distributions.
+"""
+Base.@kwdef struct MicrophysicsModel{FT <: AbstractFloat, SM}
+    "enviromental cloud fraction"
+    cf_initial::FT
+    "Subdomain statistical model"
+    statistical_model::SM
+end
+
+"""
+    MicrophysicsModel(FT;
+        cf_initial = FT(0.0),statistical_model = SubdomainMean(),)
+Constructor for `MicrophysicsModel` for EDMF, given:
+ - `FT`, the AbstractFloat type used
+ - `cf_initial`, the initial cloud fraction
+ - `statistical_model`, the assumed environmental distribution
+            of thermodynamic variables.
+"""
+function MicrophysicsModel(
+    FT;
+    cf_initial = FT(0.0),
+    statistical_model = SubdomainMean(),
+)
+    args = (cf_initial, statistical_model)
+    return MicrophysicsModel{FT, typeof(statistical_model)}(args...)
+end
+
+"""
+    Environment <: BalanceLaw
+A `BalanceLaw` for the environment subdomain arising in EDMF.
+"""
+Base.@kwdef struct Environment{FT <: AbstractFloat, N_quad} <: BalanceLaw end
+
+"""
+    Updraft <: BalanceLaw
+A `BalanceLaw` for the updraft subdomains arising in EDMF.
+"""
+Base.@kwdef struct Updraft{FT <: AbstractFloat} <: BalanceLaw end
+
+"""
+    EDMF <: TurbulenceConvectionModel
+A `TurbulenceConvectionModel` for the EDMF scheme, containing all closure models
+and free parameters.
+"""
+Base.@kwdef struct EDMF{FT <: AbstractFloat, N, UP, EN, ED, P, S, MP, ML} <:
+                   TurbulenceConvectionModel
+    "Updrafts"
+    updraft::UP
+    "Environment"
+    environment::EN
+    "Entrainment-Detrainment model"
+    entr_detr::ED
+    "Pressure model"
+    pressure::P
+    "Surface model"
+    surface::S
+    "Microphysics model"
+    micro_phys::MP
+    "Mixing length model"
+    mix_len::ML
+end
+
+"""
+    EDMF(
+        FT, N_up, N_quad;
+        updraft = ntuple(i -> Updraft{FT}(), N_up),
+        environment = Environment{FT, N_quad}(), entr_detr = EntrainmentDetrainment{FT}(),
+        pressure = PressureModel{FT}(), surface = SurfaceModel{FT}(N_up),
+        micro_phys = MicrophysicsModel(FT), mix_len = MixingLengthModel{FT}(),
+    )
+Constructor for `EDMF` subgrid-scale scheme, given:
+ - `FT`, the AbstractFloat type used
+ - `N_up`, the number of updrafts
+ - `N_quad`, the number of quadrature points used for
+        environmental distributions
```

Comment:
Is it the total number of quadrature points? Or number in one dimension? i.e. if I say `N_quad = 3` does it mean I sample my 2D histogram over 3 or 9 points?

Reply:
@yairchn ?

Reply:
this is the quadrature order (N_quad^2 is the total number), I will change the text to reflect that - thanks

Reply:
Done